### PR TITLE
WIP: Allow different mu's for different flowlines

### PR DESCRIPTION
**Pull request by fmaussion**
- [x] Fixes https://github.com/OGGM/oggm/issues/478
 - [x] Tests added/passed
 - [x] Fully documented, including `whats-new.rst` for all changes

This is a longstandig issue and not an easy one. 

I *think* I've managed to find an elegant way to deal with the calibration: it uses recursive calls to the calibration to calibrate the flowline's mu* until the entire glacier is done.

This still needs more test, and the mass-balance models need an update, hence still a WIP

### CODE REVIEW

**Comment by matthiasdusch** [comment]
The recursive approach is indeed elegant.

I have one comment/question regarding readability and also performance:
With this commit, only negative-flux-flowlines will get a different `mu_star`. So if all tributary flowlines are good, it should be possible to immediately use the already calculated `mu_star_glacierwide`. The `_mu_star_per_minimization` would then only be called on recursive calls if one or more flowlines need corrections.

But I might miss a point here or you have already some future adaptions in mind.

**Comment by fmaussion** [comment]
Thanks for looking into it!

Yes, there are some issues regarding performance here. Even the call to "minimize" is superfluous altogether since the problem is linear in mu anyway. 

However, I would like to avoid doing premature optimizations, following the good old adage*. They add complexity (here, at least one "if" and probably more), and we don't know yet if it's so slow. 



*: "The First Rule of Program Optimization: Don’t do it. 
The Second Rule of Program Optimization (for experts only!): Don’t do it yet"

**Comment by fmaussion** [comment]
This is slowly taking shape - I have added a new class to the mass-balance models which handles multiple flowlines. I spent the entire day on it - not because it's a lot of code (it's not that much code at all :D) but because it was quite hard to find the right API. I hope this is the right one now...

Anyways, still need to make this new MB the default in the production runs and also need to get the cross-validation working.

File: conftest.py [hunk 1]
```diff
@@ -1,6 +1,3 @@
-import matplotlib
-matplotlib.use('Agg')  # noqa: E402
```
Comment (matthiasdusch) [inline conftest.py]: shouldn't  this be part of PR #542 ?

Comment (fmaussion) [reply conftest.py]: ups. Yes.

File: oggm/core/massbalance.py [hunk 2]
```diff
@@ -785,7 +832,12 @@ def get_annual_mb(self, heights, year=None, fl_id=None):
 
         return self.mb_models[fl_id].get_annual_mb(heights, year=year)
 
-    def get_specific_mb(self, fls=None, year=None):
+    def get_specific_mb(self, heights=None, widths=None, fls=None,
+                        year=None):
+
+        if heights is not None or widths is not None:
+            raise ValueError('`heights` and `heights` kwargs do not work with '
```
Comment (matthiasdusch) [inline oggm/core/massbalance.py]: `heights` and `widths` kwargs...

Comment (fmaussion) [reply oggm/core/massbalance.py]: thanks!

**Comment by fmaussion** [comment]
OK, the last commit is bigger than I wish it would, BUT:
- I now added an option to compute the reference t* using the multiple flowlines approach. It is much slower than using the glacier wide one for only minimal differences in the bias (not even in the t*)! So this was done for nothing, other than helping my conscience ;-). The default remains to use glacier-wide t* calibration. 
- The climate workflow has changed it's name, I'd be happy to hear about your thoughts on this
- compute_ref_tstar is now done in parallel

So, why changing the API again? Let me try to explain.

**Before**, `mu_candidates` and `t_star_from_refmb` were separate tasks. It didn't make much sense, because they always needed to be called together:  `t_star_from_refmb` needs `mu_candidates`  and `mu_candidates`  _without_ `t_star_from_refmb` is useless.

**Now**, the mu candidates are computed and used in the same task (which kept its name): `t_star_from_refmb` . The `mu_candidates` task is still available for test and documentation purposes, but it is not used in a normal workflow. To make sure there is no confusion, I renamed it to `glacier_mu_candidates` (could be another name though).

Since I was changing things already, I reconsidered the names of the other climate calibration tasks:
- `local_mustar` is actually interpolating `t*` from the reference table. For convenience, it is also calculating the glacier wide mu* 
- `apparent_mb` was using the local mu* to add the mass flux to the branches for the inversion. Now, it is actually calculating the multiple flowline's mu* 

In my last commit, I renamed `local_mustar` to `local_tstar` and `apparent_mb` to `mu_star_calibration` to better reflect what they do.

Now that I think about it, I believe they could even be merged in one big single task, which does both the interpolation of t* and the mu* calibration in one step. I'll consider this option, more on this later...

**Comment by fmaussion** [comment]
@TimoRoth for some reason the graphics tests pass with 3.7 but not with 3.6 -> this is suspicious .

Any idea what's going on?

**Comment by fmaussion** [comment]
OK, Travis is green, let's merge this - sorry @OGGM/oggm-devs for all the changes. I suggest not to update until V1.1 is out (next couple of weeks), after which we will stabilize the code base (I promise ;)

**Comment by bearecinos** [comment]
Sorry again.... one question while using this` local_t_star() `

In my calving loop I do this:
```
    for gdir in gdirs:
        if gdir.terminus_type == 'Marine-terminating': ...... # a bunch of stuff then

                     # Lets compute the theoretical calving out of it
                     pre_calving = k * thick * w_depth * width / 1e9
                     gdir.inversion_calving_rate = pre_calving
                     print('pre_calving', pre_calving)

                     # Recompute all with calving
                     tasks.distribute_t_stars([gdir], minimum_mustar=0.)
                     tasks.apparent_mb(gdir)
                     tasks.prepare_for_inversion(gdir, add_debug_var=True)
                     tasks.volume_inversion(gdir, glen_a=cfg.A, fs=cfg.FS)
                     df = pd.read_csv(gdir.get_filepath('local_mustar')).iloc[0]
                     mu_star = df['mu_star']

```

Several questions: 
1. The minimum_mustar now I can set it up at the beginning with:  `cfg.PARAMS['min_mu_star'`]
2. after running this tasks.mass_conservation_inversion, instead of volume_inversion I want to get the local_mustar file in the glacier directory. Before it was a csv ... how can I read this new .json file?

**Pull request merged**